beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -262,6 +413,72 @@ impl<E: EthSpec> RangeBlockComponentsRequest<E> {
                         // delete it from the entries as we are going to make
                         // a separate attempt for those components.
                         requests.retain(|&k, _| k.peer != *peer);
+                        request_to_column_indices.retain(|&k, _| k.peer != *peer);
+                    }
+                }
+
+                Some(resp)
+            }
+            // Reuse same logic that we use for coupling data columns for now.
+            // todo(pawan): we should never get a coupling error here, so simplify this
+            // variant's handling.
+            RangeBlockDataRequest::DataColumnsFromRoot {
+                all_requests_made,
+                attempt,
+                request_to_column_indices,
+                expected_custody_columns,
+                requests,
+            } => {
+                // Do not couple until requests covering all required columns
+                // have been made
+                if !*all_requests_made {
+                    return None;
+                }
+
+                let mut data_columns = vec![];
+                let mut column_to_peer_id: HashMap<u64, PeerId> = HashMap::new();
+                for req in requests.values() {
+                    let Some(data) = req.to_finished() else {
+                        return None;
+                    };
+                    data_columns.extend(data.clone())
+                }
+
+                // An "attempt" is complete here after we have received a response for all the
+                // requests we made. i.e. `req.to_finished()` returns Some for all requests.
+                *attempt += 1;
+
+                // Note: this assumes that only 1 peer is responsible for a column
+                // with a batch.
+                for (id, columns) in request_to_column_indices.iter() {
+                    for column in columns {
+                        column_to_peer_id.insert(*column, id.peer);
+                    }
+                }
+
+                let expected_custody_columns: Vec<_> =
+                    expected_custody_columns.iter().copied().collect();
+                let resp = Self::responses_with_custody_columns(
```

Comment:
This returns a CouplingError with strong downscoring. Until tree sync we must not assume that the block peer is honest therefore we must not downscore column peers that do not return the expect data. Otherwise in mainnet an attacker could exploit this and force us to downscore and ban column peers, facilitating an eclipse attack.

https://github.com/sigp/lighthouse/blob/15df3d26e38d846848598f2d34e7d2b101c12f74/beacon_node/network/src/sync/range_sync/chain.rs#L905